### ISSUE TITLE
Incorrect Output Results for Quantization of my Yolov8n Model

### ISSUE THREAD

**Issue by BeyzaSimsekk** [ultralytics/ultralytics#14711]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I tried to post-training static quantization for my yolov8n model. But i did not get my expected result. I coded on Kaggle and when i run this part of my code:
```python
import matplotlib.pyplot as plt
import cv2 as cv

#Quantized modeli yükle
ort_session = ort.InferenceSession('yolov8n_quantized_static.onnx')

#Tüm resimler üzerinde inference yap (sonuçları görselleştir)
for idx, image_path in enumerate(image_paths):
    #Görselleri yükle
    image = cv.imread(image_path)
    input_tensor = preprocess_image(image_path)
    
    #Inference yap
    outputs = ort_session.run(None, {'images': input_tensor})
    
    #Sonuçları orjinal resim üzerinde göster
    detections = outputs[0] #Detections: [batch_size, num_boxes, 6] (x1,y1,x2,y2,confidence, class_id)
    """
        Buradaki her 6 eleman, tek bir nesne tespitini temsil eder:
            x1, y1 : tespit edilen nesnenin sol üst köşesinin koordinatları
            x2, y2 : tespit edilen nesnenin sağ alt köşesinin koordinatları
            confidence : tespit için modelin güven seviyesi
            class_id : tespit edilen nesnenin sınıf id'si
    """
    
    #Her bir tespiti ele alıp belirli koşullarda işleyip görselleştir
    for detection in detections[0]:
        if len(detection) >= 6:
            x1,y1,x2,y2,confidence,class_id = detection[:6]
            if confidence > 0.5: #daha güvenilir tahmin için
                x1 = int(x1)
                y1 = int(y1)
                x2 = int(x2)
                y2 = int(y2)

                #Deteceted box çiz
                cv.rectangle(image, (x1,y1), (x2,y2), (255,0,0), 2) #2 piksel kalınlığında kırmızı bir kutu

                #Class ve confidence değerini yaz
                label = f'{int(class_id)}: {confidence:.2f}'

                #Class ve confidence değerlerini kutunun üst kısmına yerleştir
                cv.putText(image, label, (x1,y1-10), cv.FONT_HERSHEY_SIMPLEX, 0.5, (255,0,0), 2)
     
    #Kaydet
    result_path = f'result_image_{idx+1}.jpg'
    success = cv.imwrite(result_path, image)
    if success:
        print(f'Succesfuly saved result for image {idx+1} at {result_path}')
    else:
        print(f'Failed to save result for image {idx+1}')
        
    
    #Görselleştir
    plt.figure(figsize = (15,15))
    plt.imshow(cv.cvtColor(image,cv.COLOR_BGR2RGB))
    plt.title(f'image {idx+1}')
    plt.axis('off')
    plt.show()`
```
my output looks like for example for the image when the person is in the center of the photo, my code gives me a result such as the blue tiny bounding boxes are on the left upper corner of the image not in the center and not around the person. when i searched this, i assumed that output of the model (or coordinates also may another reason) and my code may not match up. But i'm not sure this is the reason. But i couldn't find how to solve.

**Comment by Y-T-G** [comment]
You aren't scaling the box 

https://github.com/ultralytics/ultralytics/blob/673e76b86282859ead5517bd04dee896a647db93/examples/YOLOv8-ONNXRuntime/main.py#L110

**Comment by BeyzaSimsekk** [comment]
> You aren't scaling the box 
> 
> https://github.com/ultralytics/ultralytics/blob/673e76b86282859ead5517bd04dee896a647db93/examples/YOLOv8-ONNXRuntime/main.py#L110

thank you so much, may i also ask how can i do that? i am new to this subject. if you give me a basic roadmap, i would really be appreciated.

**Comment by pderrenger** [comment]
@BeyzaSimsekk you're welcome! To scale the bounding box coordinates correctly, you need to adjust them based on the dimensions of the input image. This ensures that the coordinates match the original image size. You can refer to the linked example for guidance on how to implement this. If you need further assistance, feel free to ask!

**Comment by BeyzaSimsekk** [comment]
Hi! Thank you, I tried to scale the boxes but everytime. it gave me another error. I guess i could not understand. Is yolov8n model quantized? I need to quantize by using onnx. I define the quantization operation like this:

`
import onnx
import onnxruntime as ort
from onnxruntime.quantization import CalibrationDataReader, create_calibrator, QuantFormat, QuantizationMode, QuantType, quantize_static, CalibrationMethod

#CustomCalibration Data Reader Sınıfı
class CustomCalibrationDataReader(CalibrationDataReader):
    def __init__(self, image_paths, batch_size = 1):
        self.image_paths = image_paths
        self.batch_size = batch_size
        self.preprocessed_data = [preprocess_image(image_path)[0] for image_path in image_paths]
        self.index = 0
        
    def get_next(self):
        if self.index >= len(self.preprocessed_data):
            return None
        data = self.preprocessed_data[self.index: self.index + self.batch_size]
        self.index += self.batch_size
        return {"images": np.concatenate(data,axis = 0)}
    

#ONNX model dosyasının yolunu belirt
onnx_model_path = 'yolov8n.onnx'
quantized_model_path = 'yolov8n_quantized_static.onnx'

#Modeli quantize et
quantize_static(
    model_input = onnx_model_path,
    model_output = quantized_model_path,
    calibration_data_reader = CustomCalibrationDataReader(image_paths),
    quant_format = QuantFormat.QDQ, #Quantization Formatı
    activation_type = QuantType.QInt8, #Aktivasyonlar için int8 quantization işlemi
    weight_type = QuantType.QInt8, #Ağırlıklar için int8 quantization işlemi
    calibrate_method = CalibrationMethod.MinMax #veya Entropy 
)
`

before this i define a preprocess operation (do i need to define my own preprocess function? if not, how can i operate?):

`
#PTQ-Static için Calibration Dataset oluştur
import cv2 as cv
import numpy as np

#Resim dosyalarının yolları
image_paths = [
    #'/kaggle/input/human-images/man_collection_crowds_event_mass_collection_group_of_people-563720.jpg',
    #'/kaggle/input/human-images/professional-profile-pictures-5760-x-3840-jrnunhjatilqof5b.jpg',
    #'/kaggle/input/human-images/164ca59a63ba1e4.jpg',
    '/kaggle/input/human-images/0422_youngpros.jpg',
    #'/kaggle/input/human-images/corporate-shake-hands.jpg',
    #'/kaggle/input/complicated-images/02acbfd6863434ec429d734f78acf1fe.jpeg',
    #'/kaggle/input/complicated-images/1684346724_mur-mur-top-p-detskii-den-rozhdeniya-kartinki-krasivo-18.jpg',
    #'/kaggle/input/complicated-images/1611889309_247232.jpg',
    #'/kaggle/input/complicated-images/7d8f8616d6a72e61545d07857ce129c2.jpg'
]

#Resimleri yüklemek ve ön işleme için fonksiyon oluştur
def preprocess_image(image_path, input_size = (640,640) ):
    image = cv.imread(image_path)
    original_height, original_width = image[:2]
    resized_image = cv.resize(image, input_size)                              #Boyutlandır
    rgb_image = cv.cvtColor(resized_image, cv.COLOR_BGR2RGB)                  #RGB'ye çevir
    normalized_image = rgb_image / 255.0                                      #0-255 aralığını 0-1 aralığına çevir
    chw_image = np.transpose(normalized_image, (2,0,1))                       #HWC'den CHW'ye (PyTorch formatı)
    input_tensor = np.expand_dims(chw_image, axis = 0).astype(np.float32)     #Batch dimension ekle
    ''' 
        Buradaki float32 modelin girişinin genellikle float32 olmasından kaynaklı bu şekilde belirtiliyor.
        Bu kodda 'input_tensor' veriyi float32 olarak döndürür çünkü bu, ONNX modelinin giriş türüyle uyumludur.
        Girişler genellikle float32 olarak sağlanır, modelin içinde quantize edilir(int8)
    '''
    return input_tensor, original_height, original_width

#Calibration Dataset Oluştur
calibration_data = [preprocess_image(image_path)[0] for image_path in image_paths]

`

but when i run this part(which i changed this part according the outputs of the model) now it doesn't show any bounding box:
`
import matplotlib.pyplot as plt
import cv2 as cv
import numpy as np
import onnxruntime as ort

#Quantized modeli yükle
ort_session = ort.InferenceSession('yolov8n_quantized_static.onnx')

#Tüm resimler üzerinde inference yap (sonuçları görselleştir)
for idx, image_path in enumerate(image_paths):
    #Görselleri yükle
    image = cv.imread(image_path)
    input_tensor, original_height, original_width = preprocess_image(image_path)
    
    #Inference yap
    outputs = ort_session.run(None, {'images': input_tensor})
    
    #print(outputs)
    
    #Sonuçları orjinal resim üzerinde göster
    detections = outputs[0] #Detections: [batch_size, num_classes, num_boxes*7] 
    
    #Detections matrisini yeniden şekillendir
    batch_size, num_classes, num_boxes_x_7 = detections.shape  
    num_boxes = num_boxes_x_7 // 7
    detections = detections.reshape(batch_size, num_classes, num_boxes, 7)
    
    # Her bir tespiti ele alıp belirli koşullarda işleyip görselleştir
    for class_id in range(detections.shape[0]):
        for box in detections[class_id]:
            if len(box) == 7:  # 7 eleman varsa
                x1, y1, x2, y2, confidence, class_confidence, _ = box
                if isinstance(confidence, np.ndarray):
                    confidence = confidence.item()  # Skalar değere dönüştür
            if confidence > 0.5:  # Daha güvenilir tahmin için
                x1 = int(x1 * original_width / 640)
                y1 = int(y1 * original_height / 640)
                x2 = int(x2 * original_width / 640)
                y2 = int(y2 * original_height / 640)

                #Detected box çiz
                cv.rectangle(image, (x1,y1), (x2,y2), (255,0,0), 2) #2 piksel kalınlığında kırmızı bir kutu

                #Class ve confidence değerini yaz
                label = f'{int(class_id)}: {confidence:.2f}'

                #Class ve confidence değerlerini kutunun üst kısmına yerleştir
                cv.putText(image, label, (x1,y1-10), cv.FONT_HERSHEY_SIMPLEX, 0.5, (255,0,0), 2)
                
        #Kaydet
        result_path = f'result_image_{idx+1}.jpg'
        success = cv.imwrite(result_path, image)
        if success:
            print(f'Succesfuly saved result for image {idx+1} at {result_path}')
        else:
            print(f'Failed to save result for image {idx+1}')
        
    
        #Görselleştir
        plt.figure(figsize = (7,7))
        plt.imshow(cv.cvtColor(image,cv.COLOR_BGR2RGB))
        plt.title(f'image {idx+1}')
        plt.axis('off')
        plt.show()

`

**Comment by pderrenger** [comment]
It looks like you're on the right track with quantizing your YOLOv8n model using ONNX. However, there are a few key points to ensure your bounding boxes are correctly scaled and displayed:

1. **Preprocessing Consistency**: Ensure that the preprocessing steps during quantization and inference are consistent. The input size should be the same during both stages.

2. **Coordinate Scaling**: After obtaining the bounding box coordinates from the model, you need to scale them back to the original image dimensions. This involves multiplying the coordinates by the ratio of the original image size to the input size used during preprocessing.

3. **Model Output Shape**: Verify the shape of the model's output to ensure you're interpreting the bounding box coordinates correctly. The output format might differ based on the model and quantization process.

Here's a concise way to address these points:

- Ensure your `preprocess_image` function is consistent for both quantization and inference.
- Scale the bounding box coordinates correctly using the original image dimensions.

If the issue persists, please ensure you are using the latest versions of the Ultralytics and ONNX packages. If you need further assistance, feel free to ask!
